### PR TITLE
Feature Guided Tour: add configuration in application.yml

### PR DESCRIPTION
**Pull request by lovaiible**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] ~~Client: I updated multiple integration tests (Jest) related to the features~~
- [x] ~~Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)~~
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] ~~Client: I added multiple screenshots/screencasts of my UI changes~~
- [x] ~~Client: I translated all the newly inserted strings into German and English~~

### Motivation and Context
This PR should solve the issue https://github.com/ls1intum/Artemis/issues/894

### Description
It is now possible to define the mapping of a tutorial to a course and exercise in the `application-dev.yml` and `application-prod.yml` file. 

### Steps for Testing
1. Log in to Artemis
2. All available tutorials can be started and completed as usual
 - Course Overview: https://artemistest.ase.in.tum.de/#/overview
 - Course Exercise Overview: https://artemistest.ase.in.tum.de/#/overview/38/exercises
 - Code Editor: https://artemistest.ase.in.tum.de/#/code-editor/153955
 - Exercise Detail: https://artemistest.ase.in.tum.de/#/overview/38/exercises/987 (only if result available)
 - Modeling: https://artemistest.ase.in.tum.de/#/modeling-submission/154013

Depending on the position of the `Tutorial Course` card (e.g. if it's on the right side) it might be that the tour step is not visible. I am aware of this issue and fix this in a follow-up PR. Since the majority of students don't have more than 2 courses, this is not critical.

### Screenshots (all existing tutorials can be started with this PR)
**Course Overview**
o
**Course Exercise Overview**
e
**Online Code Editor**
ed
**Exercise Detail for Hello World** 
s
**Modeling Tour**
m

If a user is not registered for the course, or the course for a guided tutorial does not exist, then the guided tutorial is not available:
Screen Shot 2019-11-11 at 23 40 10

### CODE REVIEW

**Comment by lovaiible** [comment]
Hi @krusche @jpbernius, as discussed today in our information sharing session, could you transfer the `guided-tour` configuration of the in `application-dev.yml` to the respective config file on the test server?

File: src/main/resources/config/application-dev.yml [hunk 1]
```diff
@@ -112,3 +112,23 @@ jhipster:
             queue-size: 512
     audit-events:
         retention-period: 120 # Number of days before audit events are deleted.
+
+# Properties to be exposed on the /info management endpoint
+info:
+    guided-tour:
+        courseShortName: 'artemistutorial'
+        tours:
+            - key: 'cancel_tour'
```
Comment (krusche) [inline src/main/resources/config/application-dev.yml]: would it not make more sense to use 'cancel_tour' as key and the '' as value?

**Comment by krusche** [comment]
@lovaiible can you please resolve the conflicts?

File: src/main/resources/config/application-prod.yml [hunk 2]
```diff
@@ -131,3 +131,16 @@ jhipster:
             queue-size: 512
     audit-events:
         retention-period: 120 # Number of days before audit events are deleted.
+
+# Properties to be exposed on the /info management endpoint
```
Comment (krusche) [inline src/main/resources/config/application-prod.yml]: please add a Github documentation page where you explain the guided tutorials and how to set them up

File: src/main/resources/config/application-dev.yml (hunk 1 above)
Comment (lovaiible) [reply src/main/resources/config/application-dev.yml]: Yes, I changed the structure of the configuration

File: src/main/resources/config/application-prod.yml (hunk 2 above)
Comment (lovaiible) [reply src/main/resources/config/application-prod.yml]: @krusche I have created a Github issue for that and will add the documentation in a follow-up PR https://github.com/ls1intum/Artemis/issues/1055

**Comment by lovaiible** [comment]
> Hello @lovaiible, thanks for this PR! I found a few things in `guided-tour.service.ts` that I had a hard time following. My comments might be a bit messy, but I hope you can follow along. :)
> Thanks for considering them.

Thank you for reviewing PR. I will have a look at your feedback and try to apply it :)

**Comment by lovaiible** [comment]
Hi @jpbernius, I have applied your feedback for this PR. Thanks again for reviewing it last time. Would you have a look at my changes and approve this PR if everything's alright?

**Comment by lovaiible** [comment]
@jpbernius @ungaralex Thanks for the review! I will have a look at the mentioned items and apply the feedback.

**Comment by krusche** [comment]
please resolve the merge conflicts and make sure to follow the new angular design guidelines described in #1081

File: src/main/webapp/app/layouts/profiles/profile.service.ts [hunk 3]
```diff
@@ -6,6 +6,7 @@ import { ProfileInfo } from 'app/layouts';
 import { BehaviorSubject } from 'rxjs';
 import { map } from 'rxjs/operators';
 import { FeatureToggleService } from 'app/feature-toggle';
+import * as _ from 'lodash';
```
Comment (maxr96) [inline src/main/webapp/app/layouts/profiles/profile.service.ts]:
```suggestion
import { reduce } from 'lodash';
```

File: src/main/webapp/app/guided-tour/guided-tour.service.ts [hunk 4]
```diff
@@ -728,15 +744,14 @@ export class GuidedTourService {
...
+        if (!course || !course.exercises || !this.guidedTourMapping) {
             return null;
         }
-        const exerciseForGuidedTourExists = course.exercises.find(
-            exercise => (exercise.type === ExerciseType.PROGRAMMING && exercise.shortName === guidedTour.exerciseShortName) || exercise.title === guidedTour.exerciseShortName,
-        );
-        if (exerciseForGuidedTourExists) {
+
+        const exerciseForGuidedTour = course.exercises.find(exercise => this.isGuidedTourAvailableForExercise(exercise, guidedTour));
+        if (this.isGuidedTourAvailableForCourse(course) && exerciseForGuidedTour) {
```
Comment (maxr96) [inline src/main/webapp/app/guided-tour/guided-tour.service.ts]: I think it is better to first execute `isGuidedTourAvailableForCourse` and only then find the exercise

File: src/main/webapp/app/guided-tour/guided-tour.service.ts [hunk 5]
```diff
@@ -98,6 +110,10 @@ export class GuidedTourService {
             });
     }
 
+    private setGuidedTourCourseExerciseMapping(guidedTourMapping: GuidedTourMapping) {
+        this.guidedTourMapping = guidedTourMapping;
+    }
```
[the remark this answers is not in the document]
Comment (lovaiible) [reply src/main/webapp/app/guided-tour/guided-tour.service.ts]: I removed it. Thanks for noticing

File: src/main/webapp/app/guided-tour/guided-tour.service.ts [hunk 6]
```diff
@@ -761,13 +779,48 @@ export class GuidedTourService {
...
+     */
+    public isGuidedTourAvailableForExercise(exercise: Exercise, guidedTour?: GuidedTour): boolean {
+        if (!exercise || !this.guidedTourMapping) {
+            return false;
+        }
+
+        let exerciseMatches = false;
+        const settingsKey = guidedTour ? guidedTour.settingsKey : this.currentTour ? this.currentTour.settingsKey : '';
+
+        if (exercise.type === ExerciseType.PROGRAMMING) {
```
[the remark this answers is not in the document]
Comment (lovaiible) [reply src/main/webapp/app/guided-tour/guided-tour.service.ts]: I think it is better to keep the logic in the guided tour service in this case :)

File: src/main/webapp/app/layouts/profiles/profile.service.ts (hunk 3 above)
Comment (maxr96) [reply src/main/webapp/app/layouts/profiles/profile.service.ts]: Why not import only what you need?

**Comment by lovaiible** [comment]
Hi @jpbernius, thank you for pointing out the issues with the highlight element when an alert pops up. Since this issue is not directly linked to this PR, I have created an issue for that and will fix that in a follow-up PR: https://github.com/ls1intum/Artemis/issues/1093

@ungaralex I think that your suggestion is really cool and interesting, however I would like to implement this in a follow-up PR if necessary.

Other than that I have applied all of you feedback, @jpbernius @maxr96 @ungaralex

**Comment by lovaiible** [comment]
On the test server somehow `Exercise.PROGRAMMING` is not evaluated but a `ReferenceError` is thrown instead. I'm currently trying to reproduce this issue and fix it.
image

**Pull request merged**